crud/stats/module.lua

```diff
@@ -0,0 +1,234 @@
+--
+-- @function disable
+--
+-- @treturn[1] boolean Returns true.
+--
+function stats.disable()
+    if stats.is_enabled() then
+        stats_registry.destroy()
+        stats_registry = nil
+    end
+
+    return true
+end
+
+--- Get statistics on CRUD operations
+--
+-- @function get
+--
+-- @treturn table Statistics on CRUD operations.
+--  If statistics disabled, returns {}.
+--
+function stats.get()
+    if stats_registry ~= nil then
+        return stats_registry.get()
+    end
+
+    return {}
+end
+
+--- Wrap CRUD operation call to collect statistics
+--
+-- @function wrap
+--
+-- @tparam function func
+--  Function to wrap. Errors are caught and thrown again.
+--
+-- @tparam string op_label
+--  Label of registry collectos.
+--  Use `require('crud.common.const').OP` to pick one.
+--
+-- @return First two arguments of wrapped function output.
+--
+function stats.wrap(func, op_label)
+    dev_checks('function', 'string')
+    return function(...)
+        if not stats.is_enabled() then
+            return func(...)
+        end
+
+        local start_time = clock.monotonic()
+        local status, res, err = pcall(func, ...)
+        local finish_time = clock.monotonic()
+
+        local success = status and (err == nil)
+        local latency = finish_time - start_time
+        stats_registry.observe(op_label, success, latency)
+
+        if status == false then
+            error(res)
+        end
+
+        return res, err
+    end
+end
+
+--- Wrap pairs operation call to collect statistics
+--
+-- @function wrap_pairs
+--
+-- @tparam function func
+--  Pairs function to wrap. Errors are caught and thrown again.
+--
+-- @tparam string op_label
+--  Label of registry collectos.
+--  Use `require('crud.common.const').OP` to pick one.
+--
+-- @return Returns pairs generator.
+--
+function stats.wrap_pairs(func, op_label)
+    dev_checks('function', 'string')
+    return function(...)
+        if not stats.is_enabled() then
+            return func(...)
+        end
+
+        local start_time = clock.monotonic()
+        local status, gen, param, state = pcall(func, ...)
+        local finish_time = clock.monotonic()
+
+        local success = status
+        local latency = finish_time - start_time
+        stats_registry.observe(op_label, success, latency)
+
+        if status == false then
+            error(gen)
+        end
+
+        return gen, param, state
+    end
+end
+
+--- Callback to collect storage tuples stats (select/pairs)
+--
+-- @function update_fetch_stats
+--
+-- @tparam table storage_stats
+--  Statistics from select storage call.
+--
+--  @tfield number tuples_fetched
+--   Count of tuples fetched during storage call.
+--
+--  @tfield number tuples_lookup
+--   Count of tuples looked up on storages while collecting response.
+--
+-- @treturn boolean Returns true.
+--
+local function update_fetch_stats(storage_stats)
+    dev_checks({ tuples_fetched = 'number', tuples_lookup = 'number' })
```

Comment:
Even checks will be disabled you recreate this table for each request. Not sure it's perf critical place but it could be saved into variable.

Reply:
Optimized